asset/src/job_metric_api/api.ts

```diff
@@ -0,0 +1,208 @@
+import {
+    OperationAPI, WorkerContext, ExecutionConfig
+} from '@terascope/job-components';
+import os from 'os';
+import {
+    Gauge, Counter, Histogram, Summary
+} from 'prom-client';
+import { JobMetricAPIConfig, JobMetricsApi } from './interfaces';
+import createExporter from './exporter';
+
+export default class Metrics extends OperationAPI<JobMetricAPIConfig> {
+    metricList: any;
+    default_labels!: Record<string, string>;
+    prefix: string;
+
+    constructor(
+        workerContext: WorkerContext,
+        apiConfig: JobMetricAPIConfig,
+        executionConfig: ExecutionConfig
+
+    ) {
+        super(workerContext, apiConfig, executionConfig);
+        this.metricList = {};
+        this.default_labels = {
+            ex_id: executionConfig.ex_id,
+            job_id: executionConfig.job_id,
+            job_name: executionConfig.name,
+            name: workerContext.sysconfig.teraslice.name,
+        };
+        this.prefix = 'teraslice_job_';
+        this.setPodName();
+    }
+
+    setPodName(): void {
+        // in a pod the hostname is the pod name
+        const host = os.hostname();
+        if (host.startsWith('ts-wkr-')) {
+            this.default_labels.pod_name = host;
+        }
+    }
+
+    /**
+     * [set metric to value]
+     * @param  {string} name [metric name]
+     * @param  {Array<string>} labelValues [list of label values]
+     * @param  {number} value [metric value]
+     * @return {Promise<void>}
+     */
+    async set(name: string, labelValues: Array<string>, value: number): Promise<void> {
+        if (this.metricList[name].functions.has('set')) {
+            this.metricList[name].metric.labels(...labelValues.concat(
+                Object.values(this.default_labels)
+            )).set(value);
+        }
+    }
+
+    /**
+     * [increment metric by value]
+     * @param  {string} name [metric name]
+     * @param  {Array<string>} labelValues [list of label values]
+     * @param  {number} value [metric value, default = 1]
+     * @return {Promise<void>}
+     */
+    async inc(name: string, labelValues: Array<string>, value = 1): Promise<void> {
```

Comment:
There is no reason this function or any function in here should return a promise since nothing requires a promise or has a need for one. Using promises all creates more of a burden for the users of this library since errors are thrown differently. For example, I had to do something like:

```ts
Promise.all(msg.metrics.map(async (metric) => {
    await api.add(
        metric.name,
        metric.help,
        ['worker_id'],
        metric.type,
    );

    const labels = [
        worker_id,
    ];
    return api.inc(metric.name, labels, metric.value);
})).catch((err) => {
    // I need to handle this error otherwise I would get a unhandledRejection
    // and crash the worker
    console.error(err);
});
```